src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketRequestExecutor.java

```diff
@@ -0,0 +1,147 @@
+     * If the requested resource is paged, or the return type is generified use this method,
+     * otherwise the {@link #makeGetRequest(HttpUrl, Class)} is most likely a better choice.
+     *
+     * @param url        url to connect to
+     * @param returnType type reference used when getting generified objects (such as pages)
+     * @param <T>        type to return
+     * @return a deserialized object of type T
+     * @see #makeGetRequest(HttpUrl, Class)
+     */
+    public <T> BitbucketResponse<T> makeGetRequest(HttpUrl url, TypeReference<T> returnType) {
+        return makeGetRequest(url, in -> objectMapper.readValue(in, returnType));
+    }
+
+    /**
+     * Make a GET request to the url given. This method will add authentication headers as needed.
+     * <em>Note!</em> this method <em>cannot</em> be used to retrieve entities that makes use of
+     * generics (such as {@link BitbucketPage}) for that use {@link #makeGetRequest(HttpUrl,
+     * TypeReference)} instead.
+     *
+     * @param url        url to connect to
+     * @param returnType class of the desired return type. Do note that if the type is generified
+     *                   this method will not work
+     * @param <T>        type to return
+     * @return a deserialized object of type T
+     * @see #makeGetRequest(HttpUrl, TypeReference)
+     */
+    public <T> BitbucketResponse<T> makeGetRequest(HttpUrl url, Class<T> returnType) {
+        return makeGetRequest(url, in -> objectMapper.readValue(in, returnType));
+    }
+
+    /**
+     * Makes a POST request to the given URL with given request payload.
+     *
+     * @param url,            the URL to make the request to
+     * @param requestPayload, JSON payload which will be marshalled to send it with POST.
+     * @param returnType,     Class of expected return type
+     * @param <T>,            Type of Request payload.
+     * @param <R>,            Return type
+     * @return
+     */
+    public <T, R> BitbucketResponse<R> makePostRequest(HttpUrl url, T requestPayload, Class<R> returnType) {
+        ObjectReader<R> reader = in -> objectMapper.readValue(in, returnType);
+        return httpRequestExecutor.executePost(url, credentials, marshall(requestPayload), response ->
+                new BitbucketResponse<>(response.headers().toMultimap(), unmarshall(reader, response.body())));
+    }
+
+    private void ensureNonEmptyBody(Response response) {
+        if (response.body() == null) {
+            log.debug("Bitbucket - No content in response");
+            throw new NoContentException(
+                    "Remote side did not send a response body", response.code());
+        }
+    }
+
+    private <T> BitbucketResponse<T> makeGetRequest(HttpUrl url, ObjectReader<T> reader) {
+        return httpRequestExecutor.executeGet(url, credentials,
+                response -> {
+                    ensureNonEmptyBody(response);
```

Comment:
Why did we move it here? considering all other validations are in `HttpRequestExecutor`?

Reply:
Empty body as part of 'Get' is not an exception. The caller should decide if that is a problem like it is done here.

Reply:
So should we move it to `BitbucketClient`? Currently `BitbucketRequestExecutor` is only responsible for converting the response into a JSON, by adding it here the exception handling is now present in `BitbucketRequestExecutor`, `HttpRequestExecutor`.

Reply:
Static analysis will struggle with this, as the `unmarshall` method require a non-null argument, but it is hard to determine (for a non-human) that it will never be null.
 I think that given that they require a non-null it is better to change the parameter list to accept null, do the null check and throw a `NoContentException` rather than a NPE.

That way you get this check everywhere it is needed not just where you remembered to add it in.